shared/fs/common/path-item-action-container.js

```diff
@@ -53,62 +50,127 @@ const mapDispatchToProps = (dispatch: Dispatch, {path}: OwnProps) => ({
+type MenuItemAppender = (
+  menuItems: Array<MenuItem>,
+  stateProps: $Call<typeof mapStateToProps, TypedState>,
+  dispatchProps: $Call<typeof mapDispatchToProps, Dispatch, OwnProps>,
+  path: Types.Path
+) => any
+
+const aShowIn: MenuItemAppender = (menuItems, stateProps, dispatchProps, path) =>
+  !isMobile && stateProps._fileUIEnabled
+    ? [
+        ...menuItems,
+        {
+          title: 'Show in ' + fileUIName,
+          onClick: dispatchProps._showInFileUI,
+        },
+      ]
+    : menuItems
+
+const aCopyPath: MenuItemAppender = (menuItems, stateProps, dispatchProps, path) => [
+  ...menuItems,
+  {
     title: 'Copy path',
-    onClick: copyPath,
-  })
-  return menuItems
+    onClick: dispatchProps.copyPath,
+  },
+]
+
+const aIgnore: MenuItemAppender = (menuItems, stateProps, dispatchProps, path) =>
+  Constants.showIgnoreFolder(path, stateProps._username)
+    ? [
+        ...menuItems,
+        {
+          title: 'Ignore this folder',
+          onClick: dispatchProps.ignoreFolder,
+          subTitle: 'The folder will no longer appear in your folders list.',
+          danger: true,
+        },
+      ]
+    : menuItems
+
+const aSave: MenuItemAppender = (menuItems, stateProps, dispatchProps, path) => {
+  const pathItem = stateProps._pathItems.get(path, Constants.unknownPathItem)
+  return isMobile && pathItem.type !== 'folder' && Constants.isMedia(pathItem)
+    ? [
+        ...menuItems,
+        {
+          title: 'Save',
+          onClick: dispatchProps._saveMedia,
+        },
+      ]
+    : menuItems
+}
+
+const aShareNative: MenuItemAppender = (menuItems, stateProps, dispatchProps, path) =>
+  isMobile && stateProps._pathItems.get(path, Constants.unknownPathItem).type === 'file'
+    ? [
+        ...menuItems,
+        {
+          title: 'Send to other app',
+          onClick: dispatchProps._shareNative,
+        },
+      ]
+    : menuItems
+
+const aDownload: MenuItemAppender = (menuItems, stateProps, dispatchProps, path) =>
+  !isMobile || (isAndroid && stateProps._pathItems.get(path, Constants.unknownPathItem).type !== 'folder')
+    ? [
+        ...menuItems,
+        {
+          title: 'Download a copy',
+          onClick: dispatchProps._download,
+        },
+      ]
+    : menuItems
+
+const tlfListAppenders: Array<MenuItemAppender> = [aShowIn, aCopyPath]
+const tlfAppenders: Array<MenuItemAppender> = [aShowIn, aIgnore, aCopyPath]
+const inTlfAppenders: Array<MenuItemAppender> = [aShowIn, aSave, aShareNative, aDownload, aCopyPath]
+
+const getRootMenuItemsByAppenders = (
+  appenders: Array<MenuItemAppender>,
+  stateProps,
+  dispatchProps,
+  path: Types.Path
+) =>
+  appenders.reduce(
+    (menuItems: Array<MenuItem>, appender: MenuItemAppender) =>
+      appender(menuItems, stateProps, dispatchProps, path),
+    ([]: Array<MenuItem>)
+  )
+
+const getRootMenuItemsByPathLevel = (pathLevel: number, stateProps, dispatchProps, path: Types.Path) => {
+  switch (pathLevel) {
+    case 0:
```

Comment:
maybe you want a constant like
```
const getPathType(path): 'keybase' | 'tlfRoot' | 'tlf' | 'tlfSub'
```

then your code will be less 
```
const needFolderList = type === 'folder' && pathElements.length >= 3,
```
and instead
```
const needFolderList = pathType === 'tlfSub'
```

Reply:
Hmm I tried this but I don't think there's gonna be a difference here other than making it more clear. We still need to check if the thing is a folder.

Reply:
sure it can be 
'keybase' | 'tlfRoot' | 'tlf' | 'tlfSub' | 'file'

Reply:
Hmm I'm not sure that'd work well: 1) To know if it's a file or a folder, we'd need to consult the store `state.fs.pathItems`, and look at a `PathItem`'s `type` field. So `getPathType` would need to accept a `PathItem` arg; 2) If we are trying to make it reflect value from the `type` field, we'd need to include not only `'file'`, but also `'symlink'`. Then we can't simply use `pathType !== 'file'` determine if a thing is a folder.